shell/platform/embedder/embedder_external_texture_gl.cc

```diff
@@ -110,6 +128,49 @@ sk_sp<DlImage> EmbedderExternalTextureGL::ResolveTexture(
   return DlImage::Make(std::move(image));
 }
 
+sk_sp<DlImage> EmbedderExternalTextureGL::ResolveTextureImpeller(
+    int64_t texture_id,
+    impeller::AiksContext* aiks_context,
+    const SkISize& size) {
+  std::unique_ptr<FlutterOpenGLTexture> texture =
+      external_texture_callback_(texture_id, size.width(), size.height());
+
+  if (!texture) {
+    return nullptr;
+  }
+
+  size_t width = size.width();
+  size_t height = size.height();
+
+  impeller::TextureDescriptor desc;
+  desc.size = impeller::ISize(width, height);
+
+  if (texture->width != 0 && texture->height != 0) {
+    width = texture->width;
+    height = texture->height;
+  }
+
+  impeller::ContextGLES& context =
+      impeller::ContextGLES::Cast(*aiks_context->GetContext());
+  impeller::HandleGLES handle = context.GetReactor()->CreateHandle(
+      impeller::HandleType::kTexture, texture->target);
+  std::shared_ptr<impeller::TextureGLES> image =
+      std::make_shared<impeller::TextureGLES>(context.GetReactor(), desc,
+                                              handle);
+
+  if (!image) {
+    // In case Skia rejects the image, call the release proc so that
+    // embedders can perform collection of intermediates.
+    if (texture->destruction_callback) {
```

Comment:
@chinmaygarde  do we have an existing way to track handles / a place for me to attach to a destruction callback? Otherwise I was just going to add an optional member to texture_gles.

Reply:
This is a case of the embedder giving us a texture handle, us using it, and then asking the embedder to collect it for us when we are done.

The most accurate way to do this is using the reactor. It knows when the texture handle is no longer in use. Attaching it to TextureGLES isn't going to be safe because of the skew between the destruction of the texture (which can happen on any thread) and the collection of the handle in the reactor (which must happen on one of the GL threads).

I did something similar (but not the same) here where the ownership of the handle is transferred in. See [the wrap](https://github.com/flutter/engine/blob/a65f1d59edc618ae81e2e8ed78d59fb729291afa/impeller/toolkit/interop/impeller.cc#L530) here of an `external_handle`. You could augment reactor handles to pass in a callback instead of a known destructor by handle (which essentially calls glDeleteTextures). That way, the handle will be collected at the right time on the right thread.

Reply:
Got it, thank you!